### DISCUSSION TITLE
Common Interchange Format and Model Descriptions for ABMs & Mesa

### DISCUSSION THREAD

**Discussion by jackiekazil** [mesa/mesa#1426] [Ideas]
Describing models - I was reading this article - https://www.jasss.org/23/2/7.html 
And we present models with a readme as a standard. I am wondering if we should take a harder look at this paper and how we present as well as maybe provide suggested documentation for people building models. 

Thoughts?

**Comment by rht** [comment]
What a coincidence. During the last meeting, we were talking about a possibility of a common formal language for ABM model specification, which can then be transpiled to a target backend for execution, be it Mesa, Agents.jl, hash.ai, etc.

For comparison, in the field of quantum computing, there is [OpenQASM](https://openqasm.com/). This allows various libraries to execute the quantum circuit specified in a different library.

**Comment by tpike3** [comment]
@jackiekazil Apologies, in drafting the NUMFOCUS letter I updated the discussion title to make a clearer link to the post. I should have cleared with your first but in the interest of time took action. Just let me know if it is a problem.

**Reply by jackiekazil** [reply]
No worries. I appreciate action. -- can you post or send me a link to the letter that you are referring to?

**Comment by wang-boyu** [comment]
Are we talking about some kind of domain-specific language (DSL) for ABMs?

Example: [Agent-Based Modelling and Simulation with Domain-Specific Languages](https://link.springer.com/chapter/10.1007/978-3-030-83039-7_7) in which the authors mentioned the [ML3](https://git.informatik.uni-rostock.de/mosi/ml3) DSL and its simulator implemented in Java.

The original paper of ML3 is probably worth reading, especially the related work section: [ML3: A language for compact modeling of linked lives in computational demography](https://ieeexplore.ieee.org/abstract/document/7408382?casa_token=Wt5OuvD_WpMAAAAA:41dh_ECcVjBv1L-aL0CbJ5ecUbnZU05C_-Q-AitZAz4iv1uJiNvjcXi8DHD0Ub7sSuRPDOqr).

**Reply by rht** [reply]
IMO starting from a subset of a known language, e.g. JavaScript/TypeScript/Python, is good. But designing yet another syntax for modelling is problematic (https://xkcd.com/927/). There is already a network effect inertia for major languages. The "links" in ML3 can be readily defined using object attributes in Python.

For those who can't be bothered to get around the paywalled ML3 paper, here is a [free link](http://simulation.su/uploads/files/default/2015-warnke-steiniger-uhrmacher-klabunde-willekens.pdf) (note: this is not Sci-Hub).

**Reply by wang-boyu** [reply]
Another DSL for ABMs: [FRED Modeling Language Guide](https://docs.epistemix.com/projects/lang-guide/en/dev8/)
